src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -27,9 +27,10 @@ namespace Microsoft.CodeAnalysis;
 /// </summary>
 public partial class Solution
 {
+    private readonly object _gate = new();
 
-    // Values for all these are created on demand.
-    private ImmutableDictionary<ProjectId, Project> _projectIdToProjectMap;
+    // Values for all these are created on demand. Only access when holding _gate.
+    private readonly Dictionary<ProjectId, Project> _projectIdToProjectMap = [];
```

Comment:
since you *do* initialize this, consider having hte dict be its own lock.

Reply:
are you auditing all usages of _projectIdToProjectMap to make sure it is locked?

Reply:
Yes, I did check all five dictionaries converted, just didn't notice the TryGetValue calls on _idToSourceGeneratedDocumentMap